src/hotspot/share/utilities/vmError.cpp

```diff
@@ -1312,6 +1313,38 @@ int VMError::prepare_log_file(const char* pattern, const char* default_pattern,
   return fd;
 }
 
+// Attempt to set the state to Native in VMError.
+// By setting it to Native, it allows os::fork_and_exec to execute cmd such as jcmd %p.
+// Otherwise, it may end up with a deadlock when cmd tries to synchronize all Java threads
+// at safepoints.
+class VMErrorThreadToNativeFromVM : public StackObj {
+ private:
+   JavaThread* _thread;
+
+ public:
+  VMErrorThreadToNativeFromVM(Thread* t) : _thread(nullptr) {
+    if (t != nullptr && t->is_Java_thread() &&
+        !Threads_lock->owned_by_self()) { // VMError::controlled_crash() grabs Threads_lock.
+      _thread = JavaThread::cast(t);
+      assert(_thread == Thread::current(), "must be current thread");
+      assert(_thread->thread_state() == _thread_in_vm, "must be in VM");
+    }
+
+    if (_thread != nullptr) {
+      assert(!_thread->owns_locks(), "must release all locks when leaving VM");
```

Comment:
This can't be an assertion as the thread is not knowingly leaving the VM without first releasing locks. If it does hold locks then that could lead to additional problems and strange errors when running the external command. We have to decide whether it is safest/best to simply not transition to native if holding locks, or whether it is okay to proceed knowing that there are risks of secondary crashes, or hangs, if we do.

Reply:
In debug build, a JavaThread can't transit to Native if it owns any lock.  Even I remove the assert here, it will hit another assert later in `ThreadStateTransition::transition_from_vm`.

```
// Checks safepoint allowed and clears unhandled oops at potential safepoints.
void JavaThread::check_possible_safepoint() {
  if (_no_safepoint_count > 0) {
    print_owned_locks();
    assert(false, "Possible safepoint reached by thread that does not allow it");
  }
```

I'd like to make VMErrorThreadToNativeFromVM only change state if _thread doesn't own any mutex, but 'Thread::own_lock()` is only available in debug build. 

Only one test  ’runtime/ErrorHandling/TestOnError.java‘ will call VMError::report_and_die with Threads_lock.

Reply:
I did flag this problem originally. It is unfortunate that we can't tell if a thread holds any locks in a product build. Not sure how to deal with this.

Reply:
> In debug build, a JavaThread can't transit to Native if it owns any lock. Even I remove the assert here, it will hit another assert later in `ThreadStateTransition::transition_from_vm`.
> 
> ```
> // Checks safepoint allowed and clears unhandled oops at potential safepoints.
> void JavaThread::check_possible_safepoint() {
>   if (_no_safepoint_count > 0) {
>     print_owned_locks();
>     assert(false, "Possible safepoint reached by thread that does not allow it");
>   }
> ```
> 
> I'd like to make VMErrorThreadToNativeFromVM only change state if _thread doesn't own any mutex, but 'Thread::own_lock()` is only available in debug build.

Which may be fine, since asserts only fire in debug builds.

Reply:
We can always walk _mutex_array like we do in print_owned_locks_on_error(). Note that locks created outside mutex_init() will not be visible though. Maybe we should fix that.